[SPMD] Remove the code branches only used in SPMD mode from distributed.py

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55353 [SPMD] Remove the code branches only used in SPMD mode from distributed.py**

Remove all the code branches that will only be executed when `device_ids > 1`.

Some helper functions are also removed:
1.  `_verify_replicas_within_process` and `verify_replicas_within_process`
2. `_replicate_modules_within_process`
3. `parallel_apply`

The next step is deprecating `_module_copies` field.

Differential Revision: [D27552201](https://our.internmc.facebook.com/intern/diff/D27552201/)

Commit:
[SPMD] Remove the code branches only used in SPMD mode from distributed.py

Remove all the code branches that will only be executed when `device_ids > 1`.

Some helper functions are also removed:
1.  `_verify_replicas_within_process` and `verify_replicas_within_process`
2. `_replicate_modules_within_process`
3. `parallel_apply`

The next step is deprecating `_module_copies` field.

Differential Revision: [D27552201](https://our.internmc.facebook.com/intern/diff/D27552201/)

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
Nice work removing all the codepaths and making DDP code much simpler :) Can you double check the CI failures though?

Commit:
Update on "[SPMD] Remove the code branches only used in SPMD mode from distributed.py"


Remove all the code branches that will only be executed when `device_ids > 1`.

Some helper functions are also removed:
1.  `_verify_replicas_within_process` and `verify_replicas_within_process`
2. `_replicate_modules_within_process`
3. `parallel_apply`

The next step is deprecating `_module_copies` field.

Differential Revision: [D27552201](https://our.internmc.facebook.com/intern/diff/D27552201/)

[ghstack-poisoned]

SciPioneer commented:
> Nice work removing all the codepaths and making DDP code much simpler :) Can you double check the CI failures though?

Last time I double checked the failures, and did not find any relevant ones. Rebased and run the tests again.

Commit:
Update on "[SPMD] Remove the code branches only used in SPMD mode from distributed.py"


Remove all the code branches that will only be executed when `device_ids > 1`.

Some helper functions are also removed:
1.  `_verify_replicas_within_process` and `verify_replicas_within_process`
2. `_replicate_modules_within_process`
3. `parallel_apply`

The next step is deprecating `_module_copies` field.

Differential Revision: [D27552201](https://our.internmc.facebook.com/intern/diff/D27552201/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Remove the code branches only used in SPMD mode from distributed.py"


Remove all the code branches that will only be executed when `device_ids > 1`.

Some helper functions are also removed:
1.  `_verify_replicas_within_process` and `verify_replicas_within_process`
2. `_replicate_modules_within_process`
3. `parallel_apply`

The next step is deprecating `_module_copies` field.

Differential Revision: [D27552201](https://our.internmc.facebook.com/intern/diff/D27552201/)

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -480,16 +476,12 @@ def __init__(
+        self._module_copies = [self.module]
         # Build parameters for reducer.
         parameters, expect_sparse_gradient = self._build_params_for_reducer()
-        # Verify model equivalence.
```
Keep the relevant part of these comments (we still need to verify model equivalence across processes)?

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -570,91 +562,26 @@ def __setstate__(self, state):
-                        and f"{module_name}.{param_name}"
-                        not in self.parameters_to_ignore
-                    ]
+        modules_and_parameters = [
```
This outermost list is always one element now? I guess to flatten in we'd need to change the reducer signature though?

SciPioneer replied:
As the description of this PR says, I plan to deprecate` _module_copies` field, which is the root cause of the extra complexity here. Will simplify it in the next PR.

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -1330,9 +1204,7 @@ def _passing_sync_batchnorm_handle(self, module_copies):
-                    layer._specify_ddp_gpu_num(
-                        len(self.device_ids) if self.device_ids else 1
-                    )
+                    layer._specify_ddp_gpu_num(1)
```
I wonder if we can always hardcode this to 1 now in sync batchnorm. Looking at the SyncBN implementation, IIUC it looks like setting this attribute is what lets the layer know we are training in DDP (which isn't very robust either since anyone can set it it seems like). So we would need to figure out some way to ensure that SyncBN is only used with DDP if we decide that we would like to remove these calls.

SciPioneer replied:
Thanks for the suggestion! Plan to remove `_specify_ddp_gpu_num` method later, as it is also used by fairscale.

SciPioneer replied:
Resolved by #56425 and #55946.

rohan-varma reviewed (approved):
LGTM, tests look clear. Please see a couple comments

Commit:
Update on "[SPMD] Remove the code branches only used in SPMD mode from distributed.py"


Remove all the code branches that will only be executed when `device_ids > 1`.

Some helper functions are also removed:
1.  `_verify_replicas_within_process` and `verify_replicas_within_process`
2. `_replicate_modules_within_process`
3. `parallel_apply`

The next step is deprecating `_module_copies` field.

Differential Revision: [D27552201](https://our.internmc.facebook.com/intern/diff/D27552201/)

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -568,93 +560,29 @@ def __setstate__(self, state):
         self.__dict__.setdefault("require_forward_param_sync", True)
         self.__dict__.setdefault("require_backward_grad_sync", True)
         parameters, expect_sparse_gradient = self._build_params_for_reducer()
+        # Verify model equivalence.
```
I think this doc should go before `dist._verify_model_across_ranks` a few lines above

SciPioneer replied:
Thanks for catching this! Didn't realize that how this comment went to that line.

Corrected in #55877
